micronota/bfillings/diamond.py

```diff
@@ -315,12 +330,105 @@ def parse_tabular(diamond_res, column='bitscore'):
         # pick the rows that have highest bitscore for each qseqid
         # df_max = df.groupby('qseqid').apply(
         #     lambda r: r[r[column] == r[column].max()])
+        idx = df.groupby('qseqid')[column].idxmax()
+        df_max = df.loc[idx]
+        df_max.index = idx.index
+        return df_max[['sseqid', 'evalue', 'bitscore']]
+
+    @staticmethod
+    def parse_sam(diamond_res, column=None, collapse=False):
+        '''Parse the output of diamond blastp/blastx.
+
+        Parameters
+        ----------
+        diamond_res : str
+            file path
+        column : str
+            The column used to pick the best hits.
+
+        Returns
+        -------
+        pandas.DataFrame
+            The best matched records for each query sequence.
+        '''
+        seqs = list(read(diamond_res, format='sam'))
+        columns = ['qseqid', 'sseqid', 'pident', 'length', 'mismatch',
+                   'gapopen', 'qstart', 'qend', 'sstart', 'send',
+                   'evalue', 'bitscore', 'sequence']
+        df = pd.DataFrame(columns=columns)
+        for i, seq in enumerate(seqs):
+            s = str(seq)
+
+            qseqid    = seq.metadata['QNAME']
+            sseqid    = seq.metadata['RNAME']
+            pident    = seq.metadata['ZI']
+            length    = seq.metadata['ZL']
+            mismatch  = seq.metadata['CIGAR']
+            gapopen   = ''
+            qstart    = seq.metadata['POS']
+            qend      = ''
+            sstart    = seq.metadata['ZS']
+            send      = ''
+            evalue    = seq.metadata['ZE']
+            bitscore  = seq.metadata['ZR']
+            row = pd.Series([qseqid, sseqid, pident,
+                             length, mismatch, gapopen,
+                             qstart, qend, sstart, send,
+                             evalue, bitscore, s],
+                            index=columns)
+            df.loc[i] = row
+
         if column is not None:
             idx = df.groupby('qseqid')[column].idxmax()
             df_max = df.loc[idx]
             df_max.index = idx.index
-            df = df_max[['sseqid', 'evalue', 'bitscore']]
+            df = df_max[['sseqid', 'evalue', 'bitscore', 'sequence']]
         else:
-            df = df[['sseqid', 'evalue', 'bitscore']]
-
+            df = df[['sseqid', 'evalue', 'bitscore','sequence']]
         return df
+
+
+class DiamondCache():
+    '''
+    Attributes
+    ----------
+    cache : list of skbio.Sequence
+        List of sequences to build cache from
+
+    '''
+    def __init__(self, seqs=None, maxSize=200000, out_dir=""):
+        self.fname = self._generate_random_file() # substitute for tempfile
```

Comment:
Here's why I did this before: http://stackoverflow.com/questions/15169101/how-to-create-a-temporary-file-that-can-be-read-by-a-subprocess

I was running into some issues with tempFile.